crates/red_knot_python_semantic/src/types/signatures.rs

```diff
@@ -11,48 +11,133 @@
 /// [`Signature`] for each overload.
 #[derive(Clone, Debug, PartialEq, Eq, Hash, salsa::Update)]
-pub enum CallableSignature<'db> {
-    Single(Signature<'db>),
-    Overloaded(Box<[Signature<'db>]>),
+pub(crate) struct CallableSignature<'db> {
+    /// The type that is (hopefully) callable.
+    pub(crate) ty: Type<'db>,
+
+    /// For an object that's callable via a `__call__` method, the type of that method. For an
+    /// object that is directly callable, the type of the object.
+    pub(crate) signature_ty: Type<'db>,
+
+    /// If this is a callable object (i.e. called via a `__call__` method), the boundness of
+    /// that call method.
+    pub(crate) dunder_call_boundness: Option<Boundness>,
+
+    /// The type of the bound `self` or `cls` parameter if this signature is for a bound method.
+    pub(crate) bound_type: Option<Type<'db>>,
+
+    overloads: Vec<Signature<'db>>,
 }
 
 impl<'db> CallableSignature<'db> {
-    /// Creates a new `CallableSignature` from an non-empty iterator of [`Signature`]s.
-    /// Panics if the iterator is empty.
-    pub(crate) fn from_overloads<I>(overloads: I) -> Self
-    where
-        I: IntoIterator,
-        I::IntoIter: Iterator<Item = Signature<'db>>,
-    {
-        let mut iter = overloads.into_iter();
-        let first_overload = iter.next().expect("overloads should not be empty");
-        let Some(second_overload) = iter.next() else {
-            return CallableSignature::Single(first_overload);
-        };
-        let mut overloads = vec![first_overload, second_overload];
-        overloads.extend(iter);
-        CallableSignature::Overloaded(overloads.into())
+    pub(crate) fn not_callable(ty: Type<'db>, signature_ty: Type<'db>) -> Self {
+        Self {
+            ty,
+            signature_ty,
+            dunder_call_boundness: None,
+            bound_type: None,
+            overloads: vec![],
+        }
     }
 
-    /// Returns the [`Signature`] if this is a non-overloaded callable, [None] otherwise.
-    pub(crate) fn as_single(&self) -> Option<&Signature<'db>> {
-        match self {
-            CallableSignature::Single(signature) => Some(signature),
-            CallableSignature::Overloaded(_) => None,
+    pub(crate) fn single(
+        ty: Type<'db>,
+        signature_ty: Type<'db>,
+        signature: Signature<'db>,
+    ) -> Self {
+        Self {
+            ty,
+            signature_ty,
+            dunder_call_boundness: None,
+            bound_type: None,
+            overloads: vec![signature],
         }
     }
 
-    pub(crate) fn iter(&self) -> std::slice::Iter<Signature<'db>> {
-        match self {
-            CallableSignature::Single(signature) => std::slice::from_ref(signature).iter(),
-            CallableSignature::Overloaded(signatures) => signatures.iter(),
+    /// Creates a new `CallableSignature` from a non-empty iterator of [`Signature`]s. Panics if
+    /// the iterator is empty.
```

Comment:
Does this still panic if the iterator is empty?

Reply:
It doesn't!  Good catch